### PR TITLE
Optim wip - Move & restructure loss objectives

### PR DESCRIPTION
**Pull request by ProGamerGov**
I moved all loss objectives from `captum.optim._core.objectives` to  `captum.optim._core.loss` as per: https://github.com/pytorch/captum/pull/500#discussion_r515453337

* I changed the loss functions to classes and made it so that they were building off an abstract loss class. The new classes work with the existing objective optimization code, so no changes to that were needed. 

* The tutorials were also updated to reflect the loss changes.

* The loss changes should be compatible with / not be in conflict with: https://github.com/pytorch/captum/pull/521

* Added a ton of missing type hints.

* I added the 3 Direction objectives that we were missing. They've been tested and work with the current objective system.

* Added Channel Interpolation objective, and made it possible to use without selecting specific channels.

* Added layer objective.

* Added alignment objective.

* Added weight objective (all 3 weight objectives merged into a single objective class).

* I got LaplacianImage working. It seems to work pretty good for feature visualization!

* I added an init parameter to FFTImage, LaplacianImage, and NaturalImage.

* FFTImage, NaturalImage, PixelImage, LaplacianImage, InputOptimization, and ToRGB all support batch sizes now!

* Added `decorrelate_init` & `squash_func` parameters to NaturalImage. 

* Added size property / function to ImageTensor, as size is more PyTorch-like than shape I think.

* Fixed major bug where errors / interrupts to the InputOptimization class' optimize function resulted in activations being None in subsequent runs.

* Custom input images now also work correctly!

test_c5

### CODE REVIEW

**Comment by ProGamerGov** [comment]
@NarineK I left the loss objective functionality the exact same, so that @greentfrapp and I can change them as we tackle issues like: https://github.com/pytorch/captum/pull/496#discussion_r509809854, and allowing for non mean & penalty losses.

**Comment by ProGamerGov** [comment]
@NarineK I don't have the required permissions to rerun the `ci/circleci: lint_test_py36_pip`, but it failed because the pip install code timed out and not because of anything related to my code.

**Review by NarineK — COMMENTED** [review]
Thank you for moving the losses into a separate file and extending from the Loss ABC,  @ProGamerGov! It looks like, there are a couple of losses that assume NCHW input shape - we could group them together or leave as is. It's probably fine to leave them as is right now.

In general I think that it would be great if we could document and explain the losses a bit more in detail and point out to the related papers.
I also added a couple inline comments. Let me know what you think.

File: captum/optim/_core/loss.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+        if x is None:
+            _x = W // 2
+        else:
+            assert x < W
+            _x = x
+
+        if y is None:
+            _y = H // 2
+        else:
+            assert y < W
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/loss.py:35]: Oops! Yeah, it should b `H` not `W`.

File: captum/optim/_core/loss.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+
+    def __call__(self, targets_to_values: ModuleOutputMapping) -> torch.Tensor:
+        activations = targets_to_values[self.target]
+        return activations ** 2
+
+
+class TotalVariation(Loss):
+    """
+    Total variation denoising penalty for activations.
+    See Simonyan, et al., 2014.
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/loss.py:112]: Lucid has the same citation, so I've reached out to Chris as it definitely looks a mistake.

File: captum/optim/_core/loss.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+                    ]
+                )
+                for j in range(activations.size(0))
+            ]
+        ) / activations.size(0)
+
+
+class ActivationInterpolation(Loss):
+    """
+    Interpolate between two different layers & channels
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/loss.py:184]: I've added a link and citiation for Activation Interpolation!

File: captum/optim/_core/loss.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+        sum_tensor = torch.zeros(1, device=activations.device)
+        for d in [1, 2, 3, 4]:
+            for i in range(B - d):
+                a, b = i, i + d
+                activ_a, activ_b = activations[a], activations[b]
+                sum_tensor = sum_tensor + (
+                    (activ_a - activ_b) ** 2
+                ).mean() / self.decay_ratio ** float(d)
+
+        return sum_tensor
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/loss.py:252]: Currently the InputOptimization class doesn't support losses like Alignment. So, I've left them as they are for @greentfrapp to change, as he's currently working on the optimization system.

File: captum/optim/_param/image/transform.py [hunk 5]
```diff
@@ -86,10 +89,15 @@ def forward(self, x, inverse=False):
         h, w = x.size("H"), x.size("W")
         flat = x.flatten(("H", "W"), "spatials")
         if inverse:
-            correct = self.transform.t() @ flat
+            correct = torch.inverse(self.transform) @ flat
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_param/image/transform.py:92]: The `self.transform `variable is actually defined as a buffer using https://pytorch.org/docs/1.1.0/nn.html#torch.nn.Module.register_buffer in ToRGB's `__init__` function:

```
    def __init__(self, transform_name: str = "klt") -> None:
        super().__init__()

        if transform_name == "klt":
            self.register_buffer("transform", ToRGB.klt_transform())
        elif transform_name == "i1i2i3":
            self.register_buffer("transform", ToRGB.i1i2i3_transform())
```

I actually made the same mistake as you when initially when first working on the ToRGB class and thought Ludwig forgot to define `self.transform`.

File: captum/optim/_core/loss.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+
+    def __call__(self, targets_to_values: ModuleOutputMapping) -> torch.Tensor:
+        activations = targets_to_values[self.target]
+
+        assert activations.dim() == 4
+
+        H_vec, W_vec = self.direction.size(2), self.direction.size(3)
+        H_activ, W_activ = activations.size(2), activations.size(3)
+
+        H = (H_activ - W_vec) // 2
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/loss.py]: I had it as `H_vec` because of the input variable name `vec`, but I've changed it to be `H_direction` now as that seems more readable.

File: captum/optim/_core/loss.py [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+    def __init__(self, target: nn.Module) -> None:
+        super(Loss, self).__init__()
+        self.target = target
+
+    @abstractmethod
+    def __call__(self, targets_to_values: ModuleOutputMapping):
+        pass
+
+    def get_neuron_pos(
+        self, H: int, W: int, x: Optional[int] = None, y: Optional[int] = None
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/loss.py]: It's specific logic for activations in the form of NCHW, and I've moved it to a new ` _utils.images.py` file.

File: captum/optim/_core/loss.py [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+        _x, _y = self.get_neuron_pos(
+            activations.size(2), activations.size(3), self.x, self.y
+        )
+        activations = activations[:, self.channel_index, _x, _y]
+        return torch.cosine_similarity(self.direction, activations[None, None, None])
+
+
+class TensorDirection(Loss):
+    """
+    Visualize a tensor direction.
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/loss.py]: I've added a citation and link to the distil paper that the direction objectives come from.

File: captum/optim/_param/image/images.py [hunk 9]
```diff
@@ -68,23 +73,27 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
...
+
+    def show(self) -> None:
         self.cpu().show()
 
-    def export(self, filename):
+    def export(self, filename: str) -> None:
         self.cpu().export(filename)
 
-    def cpu(self):
+    def cpu(self) -> "ImageTensor":
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_param/image/images.py:105]: Yes they are needed to prevent lint errors, and potentially mypy type hinting errors.

**Comment by ProGamerGov** [comment]
@NarineK I'll have to think about the ordering of the objectives, but for the moment we can leave them in their current order. 

As for the losses, I'll work on writing more in-depth descriptions with @greentfrapp, but in the meantime we can leave the descriptions as they are so that @greentfrapp can continue his work on the optimization stuff.

File: captum/optim/_core/loss.py [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+        self.direction = vec.reshape((1, -1, 1, 1))
+
+    def __call__(self, targets_to_values: ModuleOutputMapping) -> torch.Tensor:
+        activations = targets_to_values[self.target]
+        return torch.cosine_similarity(self.direction, activations)
+
+
+class DirectionNeuron(Loss):
+    """
+    Visualize a neuron direction.
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/loss.py]: I've attempted to improve the descriptions a bit, and added citations plus links for them.

File: captum/optim/_core/loss.py [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+    """
+    Total variation denoising penalty for activations.
+    See Simonyan, et al., 2014.
+    """
+
+    def _call__(self, targets_to_values: ModuleOutputMapping) -> torch.Tensor:
+        activations = targets_to_values[self.target]
+        x_diff = activations[..., 1:, :] - activations[..., :-1, :]
+        y_diff = activations[..., :, 1:] - activations[..., :, :-1]
+        return torch.sum(torch.abs(x_diff)) + torch.sum(torch.abs(y_diff))
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/loss.py:104]: It's the sum of the absolute differences for neighboring values in the activations or image. [TensorFlow's version](https://www.tensorflow.org/api_docs/python/tf/image/total_variation) links to this Wikipedia page: https://en.wikipedia.org/wiki/Total_variation_denoising

My neural-style-pt project uses basically the same algorithm, but it's origins trace back to this research article:  https://arxiv.org/abs/1412.0035

File: captum/optim/_core/loss.py [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+    Visualize a direction.
+    """
+
+    def __init__(self, target: nn.Module, vec: torch.Tensor) -> None:
+        super(Loss, self).__init__()
+        self.target = target
+        self.direction = vec.reshape((1, -1, 1, 1))
+
+    def __call__(self, targets_to_values: ModuleOutputMapping) -> torch.Tensor:
+        activations = targets_to_values[self.target]
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/loss.py:257]: I added an assert for the channel size of self.direction & activations. @greentfrapp may be able to add more assertion tests in his upcoming PR.

File: captum/optim/_core/loss.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/loss.py:97]: Okay, so according to Chris this is the correct paper: https://arxiv.org/abs/1412.0035

File: captum/optim/_core/loss.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/loss.py:105]: Chris also says that TensorFlow's total variation algorithm comes from the Understanding Deep Image Representations by Inverting Them paper: https://arxiv.org/abs/1412.0035

**Review by NarineK — APPROVED** [review]
Thank you for working on this PR, @ProGamerGov !
